### PR TITLE
Fix back-channel logouts are not sent during session termination event

### PR DESCRIPTION
**Pull request by piraveena**
Resolves https://github.com/wso2/product-is/issues/10419

Frontport the below PRs.
https://github.com/wso2-support/identity-inbound-auth-saml/pull/149
https://github.com/wso2-support/identity-inbound-auth-saml/pull/152

### CODE REVIEW

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 1]
```diff
@@ -1191,6 +1193,40 @@ private void handleAuthenticationReponseFromFramework(HttpServletRequest req, Ht
...
+     * Store SamlSSOTokenId In session context.
+     *
+     * @param samlssoTokenId       samlssoTokenId.
+     * @param authenticationResult authenticationResult.
+     */
+    private void storeSamlSSOTokenIdInContext(String samlssoTokenId, AuthenticationResult authenticationResult) {
+
+        String sessionIdentifier =
+                (String) authenticationResult.getProperty(FrameworkConstants.AnalyticsAttributes.SESSION_ID);
+        if (sessionIdentifier != null) {
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1206]: Shall we add StringUtils.isNotBlank(sessionIdentifier)

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java [hunk 2]
```diff
@@ -94,14 +95,22 @@ public String getName() {
     protected String getSamlSSOTokenIdFromEvent(Event event) {
 
         String samlssoTokenId = null;
-        if (event.getEventProperties().get(EventProperty.REQUEST) instanceof HttpServletRequest) {
+        if (event.getEventProperties().get(EventProperty.REQUEST) != null &&
+                event.getEventProperties().get(EventProperty.REQUEST) instanceof HttpServletRequest) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java:99]: let's remove redundant null check. instanceof will handle it

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java [hunk 3]
```diff
@@ -94,14 +95,22 @@ public String getName() {
...
                 if (cookie != null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Found SamlssotokenId in the request object.");
+                    }
                     samlssoTokenId = cookie.getValue();
                 }
+        }
+        if (StringUtils.isEmpty(samlssoTokenId)) {
+            if (log.isDebugEnabled()) {
+                log.debug("SamlssoTokenId is not found in the request object. Hence getting it from the context");
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java:111]: Shall we change this like "SAMLSSOTokenId". Please check in other places as well.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java [hunk 4]
```diff
@@ -94,14 +95,22 @@ public String getName() {
...
         String samlssoTokenId = null;
-        if (event.getEventProperties().get(EventProperty.REQUEST) instanceof HttpServletRequest) {
+        if (event.getEventProperties().get(EventProperty.REQUEST) != null &&
+                event.getEventProperties().get(EventProperty.REQUEST) instanceof HttpServletRequest) {
             HttpServletRequest request = (HttpServletRequest) event.getEventProperties().get(EventProperty.REQUEST);
-            if (request != null) {
                 Cookie cookie = FrameworkUtils.getCookie(request, SAMLSSOConstants.SAML_SSO_TOKEN_ID_COOKIE);
                 if (cookie != null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Found SamlssotokenId in the request object.");
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java:104]: shall we log the cookie value as well?

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java [hunk 5]
```diff
@@ -94,14 +95,22 @@ public String getName() {
...
+                    }
                     samlssoTokenId = cookie.getValue();
                 }
+        }
+        if (StringUtils.isEmpty(samlssoTokenId)) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java:109]: shall we use StringUtils.isBlank

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 6]
```diff
@@ -1191,6 +1193,40 @@ private void handleAuthenticationReponseFromFramework(HttpServletRequest req, Ht
         }
     }
 
+    /**
+     * Store SamlSSOTokenId In session context.
+     *
+     * @param samlssoTokenId       samlssoTokenId.
+     * @param authenticationResult authenticationResult.
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1200]: Start the param description with an uppercase letter and better to add it like "SAML SSO token IS" and "Authentication result". WDYT?

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java (hunk 3 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java:111]: it is a cookie name

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 6 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: changed to Authentication Result. But samlssoTokenId is a cookie name.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java [hunk 7]
```diff
@@ -96,12 +97,19 @@ protected String getSamlSSOTokenIdFromEvent(Event event) {
         String samlssoTokenId = null;
         if (event.getEventProperties().get(EventProperty.REQUEST) instanceof HttpServletRequest) {
             HttpServletRequest request = (HttpServletRequest) event.getEventProperties().get(EventProperty.REQUEST);
-            if (request != null) {
                 Cookie cookie = FrameworkUtils.getCookie(request, SAMLSSOConstants.SAML_SSO_TOKEN_ID_COOKIE);
                 if (cookie != null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Found SamlssotokenId in the request object: " + cookie.getValue());
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLLogoutHandler.java:103]:
```suggestion
                        log.debug("Found samlssoTokenId in the request object: " + cookie.getValue());
```

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 8]
```diff
@@ -1191,6 +1193,40 @@ private void handleAuthenticationReponseFromFramework(HttpServletRequest req, Ht
         }
     }
 
+    /**
+     * Store samlSSOTokenId In session context.
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1197]:
```suggestion
     * Store samlssoTokenId In session context.
```

**Comment by piraveena** [comment]
PR builder is running https://github.com/wso2/product-is/runs/1479883257

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 9]
```diff
@@ -1191,6 +1193,40 @@ private void handleAuthenticationReponseFromFramework(HttpServletRequest req, Ht
         }
     }
 
+    /**
+     * Store samlssoTokenId in session context.
+     *
+     * @param samlssoTokenId       SamlssoTokenId cookie value.
+     * @param authenticationResult Authentication Result .
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1200]: remove the excess space here.

**Review by emswbandara — APPROVED** [review]

**Review by GANGANI — APPROVED** [review]

**Pull request merged**